docs_src/design/adr/0006-Service-Registry.md

```diff
@@ -0,0 +1,106 @@
+
+ * Register with the configuration/registration (referred to simply as “the registry” for the rest of this document) provider (today Consul)
+ * Respond to shutdown requestsoTaking care to clean up of any resources in an orderly fashionoUnregister itself from the registry
+ * _Get the address to another EdgeX microservice by service name through theregistry (when enabled)_
+
+The purpose of this design is to ensure that services themselves advertise their location to the rest of the system by first self-
+registering. Most service registries (including Consul) implement some sort of health check mechanism. If a service is failing one
+or more health checks, the registry will stop reporting its availability when queried.
+
+*Note* - the design specifically excludes device services from this service lookup, as Core Metadata maintains a persistent store of
+DeviceService objects which provide service location.
+
+## Existing Behavior (Hanoi)
+
+### device-virtual-go
+I tested device-virtual-go's behavior by doing the following:
+
+$ sudo snap install edgexfoundry --edge
+$ cp /var/snap/edgexfoundry/current/config/device-virtual/res/configuration.toml .
+
+I edited the file and removed the ```[Client.Data]``` section completely and copied the
+file back into place. Next I enabled device-virtual while monitoring the journal output.
+
+$ sudo cp configuration.toml /var/snap/edgexfoundry/current/config/device-virtual/res/
+$ sudo snap set edgexfoundry device-virtual=on
+
+The following error was seen in the journal:
+
+```
+level=INFO ts=2020-11-13T23:18:17.159313493Z app=device-virtual source=httpserver.go:94 msg="Web server starting (0.0.0.0:49990)"
+error: fatal error; Host setting for Core Data client not configured
+```
+
+Next I followed the same steps, but instead of completely removing the client, I instead set the configuration items in this subsection
+to empty strings or 0 in the case of Port. The same error as above occured, and the service exited.
+
+### edgex-go
+Instead of checking every service using the same technique I used for device-virtual, I reviewed the go-mod-registry interface:
+
+```
+type Client interface {
+        // Registers the current service with Registry for discover and health check
+        Register() error
+
+        // Un-registers the current service with Registry for discover and health check
+        Unregister() error
+
+        // Simply checks if Registry is up and running at the configured URL
+        IsAlive() bool
+
+        // Gets the service endpoint information for the target ID from the Registry
+        GetServiceEndpoint(serviceId string) (types.ServiceEndpoint, error)
+
+        // Checks with the Registry if the target service is available, i.e. registered and healthy
+        IsServiceAvailable(serviceId string) (bool, error)
```

Comment:
Device services use IsServiceAvailable() on start-up to check if Core Data and Core Metadata are available, but don't request the endpoint.

Reply:
Also note that in Geneva and Hanoi the Service health checks registered by the services are used by the Consul scripts to stage gate security bootstrapping.

Reply:
I've updated the ADR to mention this, so marking this convesation resolved.